src/pytorch_lightning/strategies/tpu_spawn.py

```diff
@@ -96,17 +97,16 @@ def root_device(self) -> torch.device:
         return xm.xla_device()
 
     @staticmethod
-    def _validate_dataloader(dataloaders: Union[List[DataLoader], DataLoader]) -> None:
-        if not isinstance(dataloaders, list):
-            dataloaders = [dataloaders]
-
-        for dataloader in dataloaders:
+    def _validate_dataloader(dataloaders: Union[TRAIN_DATALOADERS, EVAL_DATALOADERS]) -> None:
+        def check_has_len(dataloader: DataLoader) -> None:
             if not has_len(dataloader):
                 raise MisconfigurationException(
                     "TPUs do not currently support IterableDataset objects, the dataset must implement `__len__`."
                     " HINT: You can mock the length on your dataset to bypass this MisconfigurationException."
                 )
 
+        apply_to_collection(dataloaders, dtype=object, wrong_dtype=(Sequence, Mapping), function=check_has_len)
```

Comment:
can you explain this change?

Reply:
Yes, I did this change to help out resolve a typing issue here. 
The entry point is  `TPUSpawnStrategy._validate_dataloader(source.instance)` we call below, and it takes `source.instance` which is a Union of a variety of collections of dataloaders. Not all these types were handled by the implementation of the validate_dataloader method previously, so I argue it's reasonable to udpate the implementation to cover all types (Sequence, Dict, ...). 

For this I used `apply_to_collection` to handle all collections. It may look strange that we use `wrong_type` here and `dtype=object`, but the intuitive solution of 
`apply_to_collection(dataloaders, dtype=DataLoader, function=check_has_len)` wouldn't work because the input can be dataloaders=DataLoader and that is not a collection.
Thus we invert the conditions.

Btw, there are tests that assert this error message is raised. We could extend the test to assert this for all supported collections.